src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -1652,30 +1660,71 @@ def height_above_floor(self, ell, e):
             sage: E.height_above_floor(2,8)
             5
         """
+        pi = self.frobenius()
+        if pi in ZZ:
+            raise ValueError("{} has a (rational) endomorphism ring of rank 4".format(self))
+
+        e = ZZ(e)
+        if not e:
+            return ZZ.zero()
+
         if self.is_supersingular():
-            raise ValueError("{} is not ordinary".format(self))
-        if e == 0:
-            return 0
+            if ell == self.base_field().characteristic():
+                # In this (exceptional) case, the Frobenius can always be divided
+                # by the maximal possible power of the characteristic. The reason
+                # is that Frobenius must be of the form phi o [p^k] where phi is
+                # a purely inseparable isogeny of degree 1 or p, hence this [p^k]
+                # can always be divided out while retaining an endomorphism.
+                assert (ell**(2*e)).divides(self.base_field().cardinality())
+                return e
+
+            # In the supersingular case, the j-invariant alone does not determine
+            # the level in the volcano. (The underlying reason is that there can
+            # be multiple non-F_q-isomorphic curves with a given j-invariant in
+            # the isogeny graph.)
+            # Example: y^2 = x^3 ± x over F_p with p congruent to 3 modulo 4 have
+            # distinct (rational) endomorphism rings.
+            # Thus we run the "probing the depths" algorithm with F_q-isomorphism
+            # classes of curves instead.
+            E0 = [self] * 3
+            E1 = [phi.codomain() for phi in self.isogenies_prime_degree(ell)]
+            assert E1
```

Comment:
I don't think that including assert statements is good practice for production code (though of course essential while testing).  Does the developers' guide say something about this?  But the alternative would be to test the condition anyway and raise an error if it fails, so perhaps it does not make much difference.  (Laos, I am sure thta my own code in Sage does include assert statements...)

Reply:
I don't think there are any rules about it. I left these in since they are very quick sanity checks that might detect if something is really wrong with the algorithm. Exceptions are for a different purpose; they communicate errors that can occur as part of the *expected* functionality of the code (e.g., invalid inputs). In my opinion, there should be assertions all over the code base anyway, with the default in production environments being that they are turned off (`python -O` or `PYTHONOPTIMIZE=1`), but that's a different discussion...